csrc/multidevice/executor.cpp

```diff
@@ -0,0 +1,140 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <device_lower/utils.h>
+#include <fusion_segmenter.h>
+#include <ir/utils.h>
+#include <multidevice/executor.h>
+
+namespace nvfuser {
+
+bool PipelineExecutor::shouldRun(PipelineStage* stage) {
+  should_run_.try_emplace(
```

Comment:
I don't think `try_emplace` here implements the desired behavior. We want to: check if `stage` is already in `should_run_`, if yes, then do nothing, else, evaluate the `std::count` and insert the result into `should_run_`. But what you are implementing here is: evaluate the `std::count`, if `stage` is already in `should_run_`, insert the count result into `should_run_`, otherwise, discard the count result.

Reply:
Maybe I'm missing something, but what's the point of using `try_emplace` instead of `emplace`?

Reply:
> I don't think `try_emplace` here implements the desired behavior. We want to: check if `stage` is already in `should_run_`, if yes, then do nothing, else, evaluate the `std::count` and insert the result into `should_run_`. 

I think on the contrary that `try_emplace` implements exactly this, which is indeed what we want... Let me quote from [here](https://en.cppreference.com/w/cpp/container/map/try_emplace): "Inserts a new element into the container with key k and value constructed with args, if there is no element with the key in the container."
Also, bottom line, I tested the code and it works... Am I missing something here? Can you explain?

> But what you are implementing here is: evaluate the `std::count`, if `stage` is already in `should_run_`, insert the count result into `should_run_`, otherwise, discard the count result.

> Maybe I'm missing something, but what's the point of using try_emplace instead of emplace?

@naoyam The point is that if the key (here "stage") is already present in the map, then the value (here the result of `count`) is discarded and not even evaluated/constructed.
This is all the more important when creating the value is costly, e.g., when we emplace the FusionExecutorCache in fec_, we save a lot of allocation/deallocation because the fusion and fec are not constructed if already cached.
The same behavior can be achieved by replacing `try_emplace` with a snippet like
```
unordered_map<key_t, value_t> my_map
if (!my_map.count(key)) {
   my_map.emplace(key, value_t(args));
}
``` 
if you prefer I can use a snippet like that instead...

Reply:
ok thx, so let me use a `if` and `emplace`